src/halo_regions.jl

```diff
@@ -1,93 +1,76 @@
-import GPUifyLoops: @launch, @loop, @unroll
-
-const PeriodicBC = BoundaryCondition(Periodic, nothing)
-const NoFluxBC = BoundaryCondition(Flux, 0)
-
-function fill_halo_regions!(grid, field_tuples...)
-    for ft in field_tuples
-        field, fbcs, data = ft
-        fill_halo_region!(grid, Val(field), fbcs, data)
-    end
+# Halo filling algorithm for arbitrary flux, gradient, and value boundary conditions.
+# For these we fill halos so that no-fluxes are added to source terms when the PDE / 
+# associated differential operators are applied on boundary elements.
+# Fluxes associated with the boundary condition are added in a separate step in the time-stepping
+# algorithm.
+ fill_west_halo!(ϕ, ::BC, H, N) = @views @. ϕ.parent[1:H, :, :] = ϕ.parent[1+H:1+H,  :, :]
+fill_south_halo!(ϕ, ::BC, H, N) = @views @. ϕ.parent[:, 1:H, :] = ϕ.parent[:, 1+H:1+H,  :]
+  fill_top_halo!(ϕ, ::BC, H, N) = @views @. ϕ.parent[:, :, 1:H] = ϕ.parent[:, :,  1+H:1+H]
+
+  fill_east_halo!(ϕ, ::BC, H, N) = @views @. ϕ.parent[N+H+1:N+2H, :, :] = ϕ.parent[N+H:N+H, :, :]
+ fill_north_halo!(ϕ, ::BC, H, N) = @views @. ϕ.parent[:, N+H+1:N+2H, :] = ϕ.parent[:, N+H:N+H, :]
+fill_bottom_halo!(ϕ, ::BC, H, N) = @views @. ϕ.parent[:, :, N+H+1:N+2H] = ϕ.parent[:, :, N+H:N+H]
```

Comment:
Hmmm, these three lines look wrong as the LHS uses `N+H+1:N+2H` which is a range of length `H-1` but the RHS uses `N+H:N+H` which is a range of length 1?

Reply:
Ah never mind, it's assigning a constant value to the entire range on the LHS.

Reply:
But then I might write `ϕ.parent[N+H, :, :]` so it doesn't look like a range on the RHS.

Reply:
This applies to all six fill functions that dispatch on `::BC`.

Reply:
It must be a range for the broadcast to work, I think. The reason is that with a single value the view is 2D; with a range (even though the range has length 1) the view is 3D. Make sense? I will add a comment.

Reply:
This is what I mean:

```julia
julia> a = rand(3, 3, 3);

julia> size(view(a, 1, :, :))
(3, 3)

julia> size(view(a, 1:1, :, :))
(1, 3, 3)
```

Reply:
Ah that makes sense. Probably obvious to a Julia veteran but otherwise yeah a comment wouldn't hurt.

Reply:
Done.